astropy/table/table.py

```diff
@@ -213,6 +216,30 @@ def _get_names_from_list_of_dict(rows):
     return list(names)
 
 
+def _get_names_from_list_of_rows(rows) -> list[str] | None:
+    """Return list of column names if ``rows`` is a list of Row objects that
+    defines table data.
+
+    If rows is not a list of Row, or if rows do not share their column names exactly,
+    then return None.
+    """
+    if rows is None or not isinstance(rows[0], Row):
+        return None
+
+    first_row = rows[0]
+    names = first_row.colnames
+    if len(rows) == 1:
+        return names
+
+    for row in rows[1:]:
+        if not isinstance(row, Row) or len(row.colnames) != len(names):
```

Comment:
note to reviewers: I'm assuming that passing rows of different sizes would raise an exception sooner or later during initialization, but I think this function should be as resilient as possible, so raising that exception is not its responsibility.